### PR TITLE
5088 manual requisition line new fields

### LINKED ISSUE #5088: New Requisition line fields
## Is your feature request related to a problem? Please describe 👀
Manual requisitions will need new fields which currently don't exist

## Describe the solution you'd like 🎁
1. Add the following fields to the Requisition line table: 
Note: All fields should be optional

| Column | Description | mSupply column |
| -- | -- | -- | 
| initial_stock_on_hand_units | Stock on hand on the date of the start of the program | Cust_prev_stock_balance |
| incoming_units | - | Cust_stock_received | 
| outgoing_units | - | Cust_stock_issued | 
| loss_in_units | Inventory Adjustment | stockLosses |
| addition_in_units | Inventory Adjustment | stockAdditions | 
| available_units | F = initial stock on hand + incoming stock + IA - outgoing stock. NOT EDITABLE | Available |
| expiring_units | sum of available stock for the item which has an expiry date after date of program | stockExpiring |
| days_out_of_stock | calculation target (Target MOS (months max stock + monthly lead time) * AMC) - available stock | days_out_or_new_demand |
| months_of_stock | - | MOS | 
| option_id | FK option table | optionID |

2. Update service + graphql requisition line endpoints with these new fields
3. Sync new fields

### Moneyworks Jobcode 🧰

OMS:CIMDV

### PR DESCRIPTION
**Pull request by Chris-Petty**
Fixes #5088

# 👩🏻‍💻 What does this PR do?

Adds fields and for manual requisitions. Sync translation updated. Basic gql added.

## 💌 Any notes for the reviewer?

A ramble about fields actually added
https://github.com/msupply-foundation/open-msupply/issues/5088#issuecomment-2425499748

> - [`days_out_of_stock` mobile uses `DOSforAMCadjustment`](https://github.com/msupply-foundation/mobile/blob/38602080d644736191062a02c024a1c7e804678c/src/sync/outgoingSyncUtils.js#L150-L151)
> - Desktop uses both `DOSforAMCadjustment` and `days_out_or_new_demand`. My suspicion here is that `days_out_or_new_demand` was the _much_ earlier field for _customer stock histories_ (basically a manual requisition) and it simply got missed? Also not ideal having a field with an "or" in its name. WHAT ARE YOU! My judgement atm is that using `days_out_or_new_demand` is correct, though perhaps at translation **push** we should populate both?
> - `Available` is not a field anywhere in OG or mobile. There is some weird circular mess that implies it's a field, but ultimately it's calculated by: `initial_SOH + incoming + additions - losses - outgoing` OG calculates this way as customisation for client - normally it's just SOH which seems bizarre. It is editable, but if you click on the value after editing it, it appears to jump back to the above custom calculation. IDK why this column is editable if calculations are going to override it (sometimes 🤪). Mobile does [something similar](https://github.com/msupply-foundation/mobile/blob/38602080d644736191062a02c024a1c7e804678c/src/database/DataTypes/RequisitionItem.js#L264-L271). I think, the code has been refactored till nothing is local to where it's used and the LSP can't help so 🤷‍♂️ can't spend too much time looking...
> - `MOS` is not a field anywhere in OG or mobile - it's just calculated from daily_usage
> - `Cust_loss_adjust` should be the net of `addition_in_units - loss_in_units` (and `initial_SOH - cust_loss_adjust >= 0` or riot) -> new field `inventory_adjustment`? Or not, because it's literally just one other field subtracted from another? We nonetheless can translate it and not add the field.
> 

# 🧪 Testing

Tests pass!

if you breakpoint OG and look at JSON pushed for a regular internal order, see the fields wow 
image
(Do we normally push `optional` fields as null?)

### CODE REVIEW

File: server/service/src/sync/translations/requisition_line.rs [hunk 1]
```diff
@@ -44,6 +44,30 @@ pub struct LegacyRequisitionLineRow {
 
     #[serde(rename = "itemName")]
     pub item_name: String,
+
+    #[serde(rename = "Cust_prev_stock_balance")]
+    pub cust_prev_stock_balance: f64,
```
Comment (roxy-dao) [inline server/service/src/sync/translations/requisition_line.rs:49]: Can you name these with the OMS names and just use serde rename to rename? Thanks

File: server/service/src/sync/translations/requisition_line.rs [hunk 2]
```diff
@@ -153,6 +193,14 @@ impl SyncTranslation for RequisitionLineTranslation {
             approved_quantity,
             approval_comment,
             item_name,
+            cust_prev_stock_balance: initial_stock_on_hand_units,
```
Comment (roxy-dao) [inline server/service/src/sync/translations/requisition_line.rs:196]: Can just use the oms name to populate

**Review by roxy-dao — CHANGES_REQUESTED** [review]
Thanks Chris just a few comments below!

> - [`days_out_of_stock` mobile uses `DOSforAMCadjustment`](https://github.com/msupply-> > > foundation/mobile/blob/38602080d644736191062a02c024a1c7e804678c/src/sync/outgoingSyncUtils.js#L150-L151)
> - Desktop uses both `DOSforAMCadjustment` and `days_out_or_new_demand`. My suspicion here is that `days_out_or_new_demand` was the _much_ earlier field for _customer stock histories_ (basically a manual requisition) and it simply got missed? Also not ideal having a field with an "or" in its name. WHAT ARE YOU! My judgement atm is that using `days_out_or_new_demand` is correct, though perhaps at translation **push** we should populate both?

If mobile is using DOSforAMCadjustment for `days_out_of_stock` then we should use translate to that field instead. I agree having a column called `days_out_or_new_demand` does cause confusion cause what even is she? Days out or new demand? 😭 

> - `Available` is not a field anywhere in OG or mobile. There is some weird circular mess that implies it's a field, but ultimately it's calculated by: `initial_SOH + incoming + additions - losses - outgoing` OG calculates this way as customisation for client - normally it's just SOH which seems bizarre. It is editable, but if you click on the value after editing it, it appears to jump back to the above custom calculation. IDK why this column is editable if calculations are going to override it (sometimes 🤪). Mobile does [something similar](https://github.com/msupply-foundation/mobile/blob/38602080d644736191062a02c024a1c7e804678c/src/database/DataTypes/RequisitionItem.js#L264-L271). I think, the code has been refactored till nothing is local to where it's used and the LSP can't help so 🤷‍♂️ can't spend too much time looking...

If this field is a customisation on a client then would it be better to create a frontend plugin for this? 

> - `MOS` is not a field anywhere in OG or mobile - it's just calculated from daily_usage

Okay that's fine, can you add all the calculations into the UI issue.

> - `Cust_loss_adjust` should be the net of `addition_in_units - loss_in_units` (and `initial_SOH - cust_loss_adjust >= 0` or riot) -> new field `inventory_adjustment`? Or not, because it's literally just one other field subtracted from another? We nonetheless can translate it and not add the field.

Yup can you add it to the push translator? Thanks

File: server/service/src/sync/translations/requisition_line.rs [hunk 3]
```diff
@@ -44,6 +44,30 @@ pub struct LegacyRequisitionLineRow {
...
+    #[serde(rename = "stockLosses")]
+    pub stock_losses: f64,
+
+    #[serde(rename = "stockAdditions")]
+    pub stock_additions: f64,
+
+    #[serde(rename = "stockExpiring")]
+    pub stock_expiring: f64,
+
+    pub days_out_or_new_demand: f64,
```
[the remark this answers is not in the document]
Comment (Chris-Petty) [reply server/service/src/sync/translations/requisition_line.rs:66]: It do tho? (I'm about to change to days_out_of_stock in the legacy row tho)
https://github.com/msupply-foundation/open-msupply/blob/b654ba6800c8b5c2f1f68d2e255fe88443dade54/server/service/src/sync/translations/requisition_line.rs#L122

[the remark this answers is not in the document]
Comment (roxy-dao) [reply server/service/src/sync/translations/requisition_line.rs:66]: OH, why did my mind just blur out the part after OR

**Comment by Chris-Petty** [comment]
> Thanks Chris just a few comments below!
> 
> > * [`days_out_of_stock` mobile uses `DOSforAMCadjustment`]([https://github.com/msupply->](https://github.com/msupply-%3E) > > foundation/mobile/blob/38602080d644736191062a02c024a1c7e804678c/src/sync/outgoingSyncUtils.js#L150-L151)
> > * Desktop uses both `DOSforAMCadjustment` and `days_out_or_new_demand`. My suspicion here is that `days_out_or_new_demand` was the _much_ earlier field for _customer stock histories_ (basically a manual requisition) and it simply got missed? Also not ideal having a field with an "or" in its name. WHAT ARE YOU! My judgement atm is that using `days_out_or_new_demand` is correct, though perhaps at translation **push** we should populate both?
> 
> If mobile is using DOSforAMCadjustment for `days_out_of_stock` then we should use translate to that field instead. I agree having a column called `days_out_or_new_demand` does cause confusion cause what even is she? Days out or new demand? 😭
> 

Nah I think that we should use `days_out_or_new_demand` because virtually all OG logic uses it including with integration with other system(s) anyway (specifically overriding `DOSforAMCadjustment` with `days_out_or_new_demand`. It's tempting to populate both on translation push though?

> > * `Available` is not a field anywhere in OG or mobile. There is some weird circular mess that implies it's a field, but ultimately it's calculated by: `initial_SOH + incoming + additions - losses - outgoing` OG calculates this way as customisation for client - normally it's just SOH which seems bizarre. It is editable, but if you click on the value after editing it, it appears to jump back to the above custom calculation. IDK why this column is editable if calculations are going to override it (sometimes 🤪). Mobile does [something similar](https://github.com/msupply-foundation/mobile/blob/38602080d644736191062a02c024a1c7e804678c/src/database/DataTypes/RequisitionItem.js#L264-L271). I think, the code has been refactored till nothing is local to where it's used and the LSP can't help so 🤷‍♂️ can't spend too much time looking...
> 
> If this field is a customisation on a client then would it be better to create a frontend plugin for this?
> 

Ugh it's worse because this field is actually ambiguous too, hinted at in the comment above it:
https://github.com/msupply-foundation/open-msupply/blob/0a6d931d9d79eefbba0b5c81b31fb4a62face189/server/service/src/sync/translations/requisition_line.rs#L27-L28

I think the SOH is also calculated the same...

https://github.com/msupply-foundation/mobile/blob/38602080d644736191062a02c024a1c7e804678c/src/database/DataTypes/RequisitionItem.js#L264-L271

https://github.com/msupply-foundation/msupply/blob/dd2166b2031f5baadd9631c5965ae292ffc8e387/Project/Sources/Methods/createIntegrationObject.4dm#L72C1-L72C190

Sooooo the `Available` column is basically just SOH? This is clapped lol. And that last snippet has it being recalculated at integration which is just... wut. 

I think that SOH is meant OK OK I SCRATCHED DEEPER:

image

image

Available literally is the `stock_on_hand` field. For a manual requisition it is the calculation as above. It probably shouldn't be editable: https://github.com/msupply-foundation/msupply/issues/15467

🤪 

> > * `MOS` is not a field anywhere in OG or mobile - it's just calculated from daily_usage
> 
> Okay that's fine, can you add all the calculations into the UI issue.
> 
> > * `Cust_loss_adjust` should be the net of `addition_in_units - loss_in_units` (and `initial_SOH - cust_loss_adjust >= 0` or riot) -> new field `inventory_adjustment`? Or not, because it's literally just one other field subtracted from another? We nonetheless can translate it and not add the field.
> 
> Yup can you add it to the push translator? Thanks

**Comment by roxy-dao** [comment]
fire

**Review by roxy-dao — APPROVED** [review]
Thankie you

**Pull request merged**